cranelift/codegen/src/isa/zkasm/inst/emit.rs

```diff
@@ -590,18 +590,15 @@ impl MachInstEmit for Inst {
                 let rd = allocs.next_writable(rd);
                 match from {
                     AMode::RegOffset(r, off, _) => {
-                        if r == context_reg() {
-                            put_string(&format!("0 => {}\n", reg_name(rd.to_reg())), sink);
-                        } else {
-                            put_string(
-                                &format!(
-                                    "$ => {} :MLOAD(MEM:{})\n",
-                                    reg_name(rd.to_reg()),
-                                    access_reg_with_offset(e0(), off)
-                                ),
-                                sink,
-                            );
-                        }
+                        debug_assert_eq!(r, e0());
```

Comment:
I don't think it is a good idea to implement this here specifically. That's because this addressing mode might very well be used e.g. for indexing an array inside a loop that's iterated by index:

```rust
for i in 0..100 {
    sum += list[i];
}
```

can become something like with loop code movement:

```rust
base_reg = list.ptr();
loop {
    if i > 400 { break; }
    sum = sum + load(base_reg + i);
    i = i + 4;
}
```

For vmctx specifically the right thing to do here would be to figure out how to make it be at a constant address that clift knows about, but that might not be trivial, since I imagine this never was a feature the design needed to support when the feature was first being implemented.

Reply:
This code should still work with the example that you provided - we still support loads and stores at addresses provided by register + offset. The crux of this PR is to catch cases when the register is a heap base pointer and substitute zero instead.
I might be missing something, though, so if you have a concrete WAT test in mind, let's add it to the test suite and check that it works.

> For vmctx specifically the right thing to do here would be to figure out how to make it be at a constant address that clift knows about, but that might not be trivial, since I imagine this never was a feature the design needed to support when the feature was first being implemented.

I agree that this would be nice, but I haven't found a way to do this yet. If you have any pointers, we can try that and revert back to using VMCTX as a base pointer instead of the custom symbol.

Reply:
Ah, alright, I might have read this part of code a 'lil too early in the day ^^